toreview/KadActionsBuilder.java

```diff
@@ -0,0 +1,311 @@
+
+    /**
+     * @param request  the received invite request
+     * @param accepted set to indicate if the answer accepts the invite or if it rejects it
+     * @return a response action built to respond to the received invite, or an invalid KAD action if the given action wasn't coherent with this building method (actionType)
+     */
+    public KadAction buildInviteAnsw(KadAction request, boolean accepted) {
+        if (request.getActionType() == KadAction.ActionType.INVITE)
+            return new KadAction(
+                    request.getPeer(),
+                    KadAction.ActionType.INVITE_ANSWER,
+                    request.getOperationId(),
+                    DEFAULT_PARTS, DEFAULT_PARTS,
+                    KadAction.PayloadType.BOOLEAN,
+                    Boolean.toString(accepted)
+            );
+        else
+            return INVALID_ACTION;
+    }
+
+
+    /**
+     * @param actionID the unique ID of the action
+     * @param peer     the receiving peer
+     * @param node     the node to be stored
+     * @return if the given ID is valid: an action built to execute the store (node) command to the peer, otherwise returns an INVALID_ACTION
+     */
+    public KadAction buildStore(int actionID, SMSPeer peer, Node<BinarySet> node) {
+        if (isValidID(actionID))
+            return new KadAction(
+                    peer,
+                    KadAction.ActionType.STORE,
+                    actionID,
+                    DEFAULT_PARTS, DEFAULT_PARTS,
+                    KadAction.PayloadType.NODE_ID,
+                    node.getKey().toHex()
+            );
+        else
+            return INVALID_ACTION;
+    }
+
+    /**
+     * @param actionID the unique ID of the action
+     * @param peer     the receiving peer
+     * @param resource the resource to be stored
+     * @return if the given ID is valid: an action built to execute the store (resource) command to the peer, otherwise returns an INVALID_ACTION
+     */
+    public KadAction buildStore(int actionID, SMSPeer peer, StringResource resource) {
+        if (isValidID(actionID))
+            return new KadAction(
+                    peer,
+                    KadAction.ActionType.STORE,
+                    actionID,
+                    DEFAULT_PARTS, DEFAULT_PARTS,
+                    KadAction.PayloadType.RESOURCE,
+                    resource.getName() + RESOURCE_SEPARATOR + resource.getValue()
+            );
+        else
+            return INVALID_ACTION;
+    }
+
+    /**
+     * @param request the received store (node) request action
+     * @param peers   the peers of which forward addresses
+     * @return an array of responses (one for each peer in peers) for the received request, or an invalid KAD action if the given action wasn't coherent with this building method (actionType)
+     */
+    public KadAction[] buildStoreAnsw(KadAction request, SMSPeer[] peers) {
+        if (request.getActionType() == KadAction.ActionType.STORE) {

```

Comment:
Vedo che ogni volta che devi costruire la risposta ad una richiesta verifichi sempre che tale richiesta corrisponda a quella che il metodo deve gestire; ovviamente questo ha un po' a che vedere con codice che non ho sotto occhio, tuttavia immagino che se viene utilizzato il metodo buildStoreAnsw**er** sia già stato verificato quale ActionType la request contenga; ha quindi senso **riverificare** che la ActionType sia quella corretta?
Comprendo comunque che debbano esserci Fail Safe, male non fanno e ci vuole poco a implementarli, ammetto di essere un po' combattuto

Reply:
si è quello che mi sono chiesto anch'io

Reply:
@turcato-niccolo inoltre come mai non sono `package-private` questi metodi? C'è una ragione di tenerli `public?`

Reply:
> Vedo che ogni volta che devi costruire la risposta ad una richiesta verifichi sempre che tale richiesta corrisponda a quella che il metodo deve gestire; ovviamente questo ha un po' a che vedere con codice che non ho sotto occhio, tuttavia immagino che se viene utilizzato il metodo buildStoreAnsw**er** sia già stato verificato quale ActionType la request contenga; ha quindi senso **riverificare** che la ActionType sia quella corretta?
Anche se l'utente utilizza solo questo builder e non la classe KadAction, potrebbe scrivere `buildStoreAnsw(buildPing(), ...)` cosa che è sbagliata su più fronti. (iperseplificazione di uno scenario di errore, ma tanto per capirci)
E' sempre la questione che è emersa ieri: gestisco la situazione di errore o dico nelle specifiche "attento che devi passarmi solo dati coerenti, se non lo fai me ne lavo le mani", è una scelta. Visto che questi metodi alla fine sono un po' poveri, ho preferito aggiungerci una logica in più